### DISCUSSION TITLE
Sobre timezone e seus bugs

### DISCUSSION THREAD

**Discussion by felquis** [frontendbr/forum#363] [Perguntas e Respostas]
Estava fazendo alguns testes em um projeto e mudei várias vezes o dia e hora do meu computador, no meio destes testes fiz um commit que deixou o GitHub meio atordoado.

Este commit https://github.com/pagarme/react-event-components/commit/d156803125775a7d518e65f5aeb5116acce12ea4 foi feito no dia 7 de novembro, porém a data do meu computador era dia 11 de novembro, e o GitHub hoje 10 de novembro as 13:12 mostra como se o arquivo foi alterado just now!

![screenshot from 2016-11-09 22-39-39](https://cloud.githubusercontent.com/assets/736728/20182015/727636a0-a747-11e6-9d14-0d0c1384c2fd.png)
> imagem 1

Claro que essa foi nossa reação 
![de_volta_para_o_futuro](https://cloud.githubusercontent.com/assets/736728/20182099/c4d1f89e-a747-11e6-9bb4-24f6e21e969c.jpeg)

O mais interessante é que na [página do commit](https://github.com/pagarme/react-event-components/commit/d156803125775a7d518e65f5aeb5116acce12ea4) com o diff diz que foi comitado um dia a partir de agora.
![screenshot from 2016-11-10 13-16-17](https://cloud.githubusercontent.com/assets/736728/20182123/e3444da4-a747-11e6-9414-de95836a2ef0.png)
> imagem 2

:beetle: a situação já é bizarra, mas temos aqui dois timers com outputs diferentes.

Na imagem 1, o elemento usado é o [<time-ago>](https://github.com/github/time-elements/blob/597103e765b9c9d4570c6f1ee651ead8017018c2/time-elements.js#L599) e na imagem 2 é usado o [<relative-time>](https://github.com/github/time-elements/blob/597103e765b9c9d4570c6f1ee651ead8017018c2/time-elements.js#L595) a diferença entre os dois basicamente é o que é usado no método [RelativeTimePrototype.getFormattedDate](https://github.com/github/time-elements/blob/597103e765b9c9d4570c6f1ee651ead8017018c2/time-elements.js#L416) e [TimeAgoPrototype.getFormattedDate](https://github.com/github/time-elements/blob/597103e765b9c9d4570c6f1ee651ead8017018c2/time-elements.js#L446)

O interessante é que no DevTools, no componente da imagem 1 mudei a tag `time-ago` para `relative-time` e o resultado foi que mostrou que o commit foi de "a day from now", um dia a partir de agora.

![screenshot from 2016-11-10 13-23-38](https://cloud.githubusercontent.com/assets/736728/20182677/ba7dcdda-a749-11e6-93b8-dc88a6aa7fdd.png)

Isso é algo adverso que aprendi esses dias que queria compartilhar com vocês :)

Foi inevitavelmente aqui no trabalho [acessar uma tabela de timezones do planeta](https://en.wikipedia.org/wiki/Time_zone#/media/File:Standard_World_Time_Zones.png) e conversar muito sobre possiveis bugs e soluções.

O que eu queria mesmo é saber de vocês coisas como:

#### Como você corrigiria este problema de horário desconfigurado do GitHub? 

Abs

**Comment by felipefialho** [comment]
Cara, sensacional como você curte bugar o Github, hahaha.

Mudei o titulo para ficar mais fácil a compreensão do assunto, beleza? 😄

**Comment by felquis** [comment]
@LFeh o bizarro é como nós devs podemos se precaver de bugs relacionado a hora do usuário estar errada, nem é o timezone em si, mas sim o computador do client estar desconfigurado.

**Comment by fdaciuk** [comment]
Trabalhar com datas / horas sempre dá essas tretas xD
O nosso calendário mesmo já um bug por si só: temos 365 dias e 6 horas por ano, onde essas 6 horas são acumuladas a cada 4 anos, que tem um dia a mais pra "tentar" fechar a conta xD

Mas o esquema da data no git não vejo que seja um bug, mas sim o comportamento esperado. O git é um sistema distribuído, então cada commit vai respeitar o horário atual da máquina do usuário mesmo. 

O GitHub mantém esse histórico do git, mas provavelmente o `time-ago` é para que ele possa ter controle do tempo, com base nos servers deles.

Mas já vi muito disso: a galera esquece de deixar o PC configurado pra acertar sozinho no horário de verão e acaba fazendo o commit uma hora adiantado :joy: :joy:

O problema é só na visualização do histórico, que va ficar meio zuado :stuck_out_tongue_winking_eye:

**Comment by fititnt** [comment]
@felquis tipo o que é citado aqui https://github.com/isaacs/github/issues/374 ?

Eu sou da opinião que implementações, quando existe algo parecido com uma _norma técnica_, só são "obrigadas" a implementar de forma esperada ações esperadas, podendo agir de forma inesperada para ações inesperadas.

Para uma ação inesperada, se eu fosse escolher implementar, eu faria o que é mais fácil e teria menos chances de bugs. Isto é, fazer todo um processo para contornar uma situação que raramente seria usada e que não tem implicações de segurança não vale a pena.
